src/eval.rs

```diff
@@ -1,26 +1,998 @@
+                    UOp::Negate => cast(-l, range, Overflow::Wraparound),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                Value::Bool(b) => match op {
+                    UOp::Not => Ok(Value::Bool(!b)),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::Cast {
+            inner,
+            typ,
+            overflow,
+        } => {
+            let cast_t = match typ {
+                Type::BoolT => Err(RuntimeError::TypeError),
+                Type::SymT => Err(RuntimeError::TypeError),
+                Type::NumT(n) => Ok(n.clone()),
+            }?;
+
+            let val = eval_expr(inner, env, &program)?;
+
+            match val {
+                Value::Num(num, range) => match overflow {
+                    Overflow::Fail => cast(num, cast_t, Overflow::Fail),
+                    Overflow::Saturate => cast(num, cast_t, Overflow::Saturate),
+                    Overflow::Wraparound => cast(num, cast_t, Overflow::Wraparound),
+                },
+                // can only cast ints
+                Value::Bool(b) => Err(RuntimeError::TypeError),
+                Value::Sym(s) => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::Match { scrutinee, cases } => {
+            let raw_val = eval_expr(&scrutinee, env, &program)?;
+            for case in cases {
+                let pattern = match case.pattern {
+                    Pattern::Bool(b) => Value::Bool(b),
+                    Pattern::Num(n) => Value::Num(
+                        n,
+                        Range {
+                            start: n,
+                            end: n + 1,
+                        },
+                    ), // TODO: how would we match on Num, not including the range?
+                    Pattern::Sym(s) => Value::Sym(s),
+                    Pattern::Var(id) => env.get(&id).unwrap().clone(),
+                };
+                if raw_val == pattern {
+                    let g = eval_expr(&case.guard, env, &program).unwrap();
+
+                    if let Value::Bool(b) = g {
+                        if b {
+                            return Ok(eval_expr(&case.result, env, &program).unwrap());
+                        }
+                    } else {
+                        return Err(RuntimeError::TypeError);
+                    }
+                }
+            }
+            return Err(RuntimeError::TypeError);
+        }
+
+        Expr::Call { callee, args } => {
+            let mut env_args = Env::new();
+            let function = program.helpers.get(callee).unwrap().clone();
+
+            // check to see proper args
+            if args.len() != function.params.len() {
+                return Err(RuntimeError::IncorrectArgs);
+            }
+            let mut num_matches = 0; // checking to see if we've matched every arg
```

Comment:
you don't need to do any runtime type checking here. any typing issues will show up later when the values are used, and we have a type checker.